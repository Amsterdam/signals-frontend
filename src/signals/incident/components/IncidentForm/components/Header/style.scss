@import "config.scss";

.header {
  &__label {
    @include avenirnextdemi();
    margin-bottom: 8px;
  }

  &__subtitle {
    color: $ams-donkerantraciet;
    font-size: 14px;
    margin-bottom: 4px;
  }

  &--not-required {
    @include avenirnext();
    margin-left: 8px;
  }

  &__errors {
    &__item {
      color: $ams-rood;
      font-size: 14px;
      margin-bottom: 4px;
    }
  }

  &--invalid {
    border-left: $ams-rood 2px solid;
    padding-left: 12px;
  }
}
