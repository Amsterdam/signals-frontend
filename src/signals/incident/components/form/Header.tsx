import styled from 'styled-components'

const Header = styled.header`
  font-weight: 700;
  font-size: 1.25rem;
`

export const QuestionHeader = ({ meta }: any) => {
  return <Header>{meta.label}</Header>
}
