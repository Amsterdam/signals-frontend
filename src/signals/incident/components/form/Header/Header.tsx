// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2021 Gemeente Amsterdam
import type { FunctionComponent } from 'react'
import React, { Fragment } from 'react'
import styled from 'styled-components'
import { themeSpacing, themeColor } from '@amsterdam/asc-ui'
import Label from 'components/Label'
import ErrorMessage, { ErrorWrapper } from 'components/ErrorMessage'
import type {
  ReactiveFormMeta,
  FormMeta,
  FormOptions,
} from 'types/reactive-form'

const StyledErrorWrapper = styled(ErrorWrapper)<{ invalid: boolean }>`
  display: flex;
  flex-direction: column;

  & > :last-child:not(& > :first-child) {
    margin-top: ${themeSpacing(3)};
  }
`

const StyledLabel = styled(Label)`
  margin-bottom: 0;
  line-height: ${themeSpacing(6)};
`

const Optional = styled.span`
  font-family: Avenir Next LT W01-Regular, arial, sans-serif;
  margin-left: ${themeSpacing(2)};
`

const SubTitle = styled.p`
  color: ${themeColor('tint', 'level5')};
  margin-top: 0;
  margin-bottom: 0;
  line-height: ${themeSpacing(6)};
`

type PickedProps = 'touched' | 'hasError' | 'getError'
export interface HeaderProps extends Pick<ReactiveFormMeta, PickedProps> {
  className?: string
  meta: FormMeta
  options?: FormOptions
}

const Header: FunctionComponent<HeaderProps> = ({
  className,
  meta,
  options,
  touched,
  hasError,
  getError,
  children,
}) => {
  const containsErrors: boolean =
    touched &&
    (hasError('required') ||
      hasError('email') ||
      hasError('maxLength') ||
      hasError('custom'))
  const isOptional = !options?.validators?.some(
    (validator) => validator.name === 'required'
  )

  return (
    <StyledErrorWrapper className={className} invalid={containsErrors}>
      {meta?.label && (
        <StyledLabel htmlFor={meta.name}>
          {meta.label}

          {isOptional && <Optional>(niet verplicht)</Optional>}
        </StyledLabel>
      )}

      {meta?.subtitle && (
        <SubTitle id={`subtitle-${meta.name}`}>{meta.subtitle}</SubTitle>
      )}

      {touched && containsErrors && (
        <Fragment>
          {hasError('required') && (
            <ErrorMessage
              data-testid={`${meta.name}-required`}
              message={
                getError('required') === true
                  ? 'Dit is een verplicht veld'
                  : (getError('required') as string)
              }
            />
          )}

          {hasError('email') && (
            <ErrorMessage
              data-testid="invalid-mail"
              message="Vul een geldig e-mailadres in, met een @ en een domeinnaam. Bijvoorbeeld: naam@domein.nl"
            />
          )}

          {hasError('maxLength') && (
            <ErrorMessage
              message={`U heeft meer dan de maximale ${String(
                (getError('maxLength') as { requiredLength: number })
                  .requiredLength
              )} tekens ingevoerd`}
            />
          )}

          {hasError('custom') && (
            <ErrorMessage message={getError('custom') as string} />
          )}
        </Fragment>
      )}

      {children}
    </StyledErrorWrapper>
  )
}

export default Header
