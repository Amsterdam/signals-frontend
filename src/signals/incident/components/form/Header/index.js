import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { themeSpacing, themeColor } from '@datapunt/asc-ui';

import { Validators } from 'react-reactive-form';

const Children = styled.div`
  display: flex;
  flex-flow: column;

  & > * + * {
    margin-top: ${themeSpacing(2)};
  }
`;

const Wrapper = styled.div`
  ${({ invalid }) =>
    invalid &&
    css`
      border-left: ${themeColor('support', 'invalid')} 2px solid;
      padding-left: ${themeSpacing(3)};
    `}
`;

const Label = styled.div`
  font-family: Avenir Next LT W01 Demi;
  margin-bottom: 10px;
`;

const Optional = styled.span`
  font-family: Avenir Next LT W01-Regular;
  margin-left: ${themeSpacing(2)};
`;

const ErrorItem = styled.div`
  color: ${themeColor('support', 'invalid')};
  font-size: 14px;
  margin-bottom: ${themeSpacing(1)};
`;

const SubTitle = styled.div`
  color: ${themeColor('tint', 'level5')};
  margin-top: ${themeSpacing(-1)};
  margin-bottom: ${themeSpacing(2)};
`;

const Header = ({ className, meta, options, touched, hasError, getError, children }) => {
  const containsErrors =
    touched && (hasError('required') || hasError('email') || hasError('maxLength') || hasError('custom'));
  const isOptional = !options?.validators?.includes(Validators.required);

  return (
    <Wrapper className={className} invalid={containsErrors}>
      {meta?.label && (
        <Label>
          {meta.label}

          {isOptional && <Optional>(optioneel)</Optional>}
        </Label>
      )}

      {meta?.subtitle && <SubTitle>{meta.subtitle}</SubTitle>}

      {touched && containsErrors && (
        <Fragment>
          {hasError('required') && <ErrorItem>Dit is een verplicht veld</ErrorItem>}

          {hasError('email') && <ErrorItem>Het moet een geldig e-mailadres zijn</ErrorItem>}

          {hasError('maxLength') && (
            <ErrorItem>U kunt maximaal {getError('maxLength').requiredLength} tekens invoeren.</ErrorItem>
          )}

          {hasError('custom') && <ErrorItem>{getError('custom')}</ErrorItem>}
        </Fragment>
      )}

      <Children>{children}</Children>
    </Wrapper>
  );
};

Header.defaultProps = {
  className: '',
};

Header.propTypes = {
  className: PropTypes.string,
  meta: PropTypes.shape({
    label: PropTypes.string,
    subtitle: PropTypes.string,
  }),
  options: PropTypes.shape({
    validators: PropTypes.arrayOf(PropTypes.any),
  }),
  touched: PropTypes.bool,
  hasError: PropTypes.func.isRequired,
  getError: PropTypes.func,
  children: PropTypes.node,
};

export default Header;
