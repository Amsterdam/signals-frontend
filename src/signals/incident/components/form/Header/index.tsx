import type { FunctionComponent } from 'react';
import React, { Fragment } from 'react';
import styled, { css } from 'styled-components';
import { themeSpacing, themeColor } from '@amsterdam/asc-ui';
import Label from 'components/Label';

const Wrapper = styled.div<{ invalid: boolean }>`
  display: flex;
  flex-direction: column;

  ${({ invalid }) =>
    invalid &&
    css`
      border-left: ${themeColor('support', 'invalid')} 2px solid;
      padding-left: ${themeSpacing(3)};
    `}

  /* keep a 3 units margin above the last element/control  */
  & > :last-child :not(& > :first-child) {
    margin-top: ${themeSpacing(3)};
  }
`;

const StyledLabel = styled(Label)`
  margin-bottom: 0;
  line-height: ${themeSpacing(6)};
`;

const Optional = styled.span`
  font-family: Avenir Next LT W01-Regular, arial, sans-serif;
  margin-left: ${themeSpacing(2)};
`;

const ErrorItem = styled.p`
  font-family: Avenir Next LT W01 Demi, arial, sans-serif;
  margin-top: 0;
  margin-bottom: 0;
  color: ${themeColor('support', 'invalid')};
  line-height: ${themeSpacing(6)};
`;

const SubTitle = styled.p`
  color: ${themeColor('tint', 'level5')};
  margin-top: 0;
  margin-bottom: 0;
  line-height: ${themeSpacing(6)};
`;

export interface HeaderProps {
  meta: {
    name: string;
    label?: string;
    subtitle?: string;
  };
  options?: {
    validators?: { name: string }[];
  };
  touched: boolean;
  hasError: (name: string) => boolean;
  getError: (name: string) => string | boolean | { requiredLength: number };
}

const Header: FunctionComponent<HeaderProps> = ({ meta, options, touched, hasError, getError, children }) => {
  const containsErrors: boolean =
    touched && (hasError('required') || hasError('email') || hasError('maxLength') || hasError('custom'));
  const isOptional = !options?.validators?.some(validator => validator.name === 'required');

  return (
    <Wrapper invalid={containsErrors}>
      {meta?.label && (
        <StyledLabel htmlFor={meta.name}>
          {meta.label}

          {isOptional && <Optional>(optioneel)</Optional>}
        </StyledLabel>
      )}

      {meta?.subtitle && <SubTitle id={`subtitle-${meta.name}`}>{meta.subtitle}</SubTitle>}

      {touched && containsErrors && (
        <Fragment>
          {hasError('required') && (
            <ErrorItem>{getError('required') === true ? 'Dit is een verplicht veld' : getError('required')}</ErrorItem>
          )}

          {hasError('email') && (
            <ErrorItem>
              Vul een geldig e-mailadres in, met een @ en een domeinnaam. Bijvoorbeeld: naam@domein.nl
            </ErrorItem>
          )}

          {hasError('maxLength') && (
            <ErrorItem>U heeft meer dan de maximale {String(getError('maxLength').requiredLength)} tekens ingevoerd</ErrorItem>
          )}

          {hasError('custom') && <ErrorItem>{getError('custom')}</ErrorItem>}
        </Fragment>
      )}

      {children}
    </Wrapper>
  );
};

export default Header;
