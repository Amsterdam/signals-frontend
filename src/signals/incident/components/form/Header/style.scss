@import "config.scss";

.header {
  &__label {
    @include avenirnextdemi();
    margin-bottom: 10px;
  }

  &__subtitle {
    color: $ams-donkergrijs;
    margin-top: -5px;
    margin-bottom: 8px;
  }

  &--not-required {
    @include avenirnext();
    margin-left: 8px;
  }

  &__errors {
    &__item {
      color: $ams-rood;
      font-size: 14px;
      margin-bottom: 4px;
    }
  }

  &--invalid {
    border-left: $ams-rood 2px solid;
    padding-left: 12px;
  }
}
