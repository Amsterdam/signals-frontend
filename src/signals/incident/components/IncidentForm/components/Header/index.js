import React from 'react';
import PropTypes from 'prop-types';

import { Validators } from 'react-reactive-form';

import './style.scss';

const Header = ({ meta, options, touched, hasError, getError, children }) => (
  <div className={`header ${touched && (hasError('required') || hasError('email') || hasError('maxLength')) ? 'header--invalid' : ''}`}>
    <div className="header__label">{meta && meta.label}
      {(meta.label && (!options || !options.validators)) || (options && options.validators && !options.validators.includes(Validators.required)) ?
        <span className="header--not-required">(niet verplicht)</span>
      : ''}
    </div>
    <div className="header__subheader">{meta && meta.subheader}</div>

    <div className="header__errors">
      <div className="header__errors__item">
        {touched
        && hasError('required')
        && 'Dit is een verplicht veld'}
      </div>

      <div className="header__errors__item">
        {touched
        && hasError('email')
        && 'Het moet een geldig e-mailadres zijn'}
      </div>

      <div className="header__errors__item">
        {touched
        && hasError('maxLength')
        && `U kunt maximaal ${getError('maxLength').requiredLength} tekens invoeren.`}
      </div>
    </div>

    <div className="header__children">
      {children}
    </div>
  </div>
);

Header.propTypes = {
  meta: PropTypes.object,
  options: PropTypes.object,
  touched: PropTypes.bool,
  hasError: PropTypes.func,
  getError: PropTypes.func,
  children: PropTypes.object
};

export default Header;
