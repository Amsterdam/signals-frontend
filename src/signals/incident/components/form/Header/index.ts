// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 Gemeente Amsterdam
import Header from './Header'

export default Header
