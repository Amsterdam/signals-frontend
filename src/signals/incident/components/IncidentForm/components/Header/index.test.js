import React from 'react';
import { shallow } from 'enzyme';

import Header from './index';

describe('Form component <Header />', () => {
  let wrapper;
  let meta;
  let options;
  let touched;
  let getError;
  let hasError;

  beforeEach(() => {
    meta = {};
    options = {};
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();

    wrapper = shallow(<Header
      meta={meta}
      options={options}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('error messages', () => {
    it('should render no error when element is only touched', () => {
      wrapper.setProps({ touched: true });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render required error', () => {
      hasError.mockImplementation((type) => type === 'required');
      wrapper.setProps({ touched: true });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render email error', () => {
      hasError.mockImplementation((type) => type === 'email');
      wrapper.setProps({ touched: true });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render maxLength error', () => {
      hasError.mockImplementation((type) => type === 'maxLength');
      getError.mockImplementation(() => ({
        requiredLength: 666
      }));
      wrapper.setProps({ touched: true });

      expect(wrapper).toMatchSnapshot();
    });
  });
});
