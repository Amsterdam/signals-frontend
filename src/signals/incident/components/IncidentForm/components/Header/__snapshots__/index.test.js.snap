// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <Header /> error messages should render email error 1`] = `
<div
  className="header header--invalid"
>
  <div
    className="header__label"
  />
  <div
    className="header__subheader"
  />
  <div
    className="header__errors"
  >
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    >
      Het moet een geldig e-mailadres zijn
    </div>
    <div
      className="header__errors__item"
    />
  </div>
  <div
    className="header__children"
  />
</div>
`;

exports[`Form component <Header /> error messages should render maxLength error 1`] = `
<div
  className="header header--invalid"
>
  <div
    className="header__label"
  />
  <div
    className="header__subheader"
  />
  <div
    className="header__errors"
  >
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    >
      Dit veld mag niet langer dan 666 tekens zijn.
    </div>
  </div>
  <div
    className="header__children"
  />
</div>
`;

exports[`Form component <Header /> error messages should render no error when element is only touched 1`] = `
<div
  className="header "
>
  <div
    className="header__label"
  />
  <div
    className="header__subheader"
  />
  <div
    className="header__errors"
  >
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
  </div>
  <div
    className="header__children"
  />
</div>
`;

exports[`Form component <Header /> error messages should render required error 1`] = `
<div
  className="header header--invalid"
>
  <div
    className="header__label"
  />
  <div
    className="header__subheader"
  />
  <div
    className="header__errors"
  >
    <div
      className="header__errors__item"
    >
      Dit is een verplicht veld
    </div>
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
  </div>
  <div
    className="header__children"
  />
</div>
`;

exports[`Form component <Header /> should render correctly 1`] = `
<div
  className="header "
>
  <div
    className="header__label"
  />
  <div
    className="header__subheader"
  />
  <div
    className="header__errors"
  >
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
    <div
      className="header__errors__item"
    />
  </div>
  <div
    className="header__children"
  />
</div>
`;
